### PR TITLE
MGMT-71 Add documentation for managers on how to coach for productivity

### PR DESCRIPTION
**Pull request by dbhynds**
## Summary of Changes

* [x] **I acknowledge that this is a public repository, visible to anyone on the internet.**

Primarily this adds a document for how to troubleshoot performance problems and help engineers grow in areas where they are struggling. This came out of a conversation with Jordan about how to know "which tool to reach for." This knowledge is something I've developed intuitively over time. As our engineering management team grows (not to mention the ICs on the team), I think it'd be helpful to get this documented somewhere.

The list is far from exhaustive, but should serve as a starting point. If you have any other ideas of common problems engineers face, let me know, and I can add them in.

### CODE REVIEW

File: productivity.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+- How does the person feel about their tickets?
+  - Can they identify any trends in the problems they encounter with tickets?
+- How do they feel asking for help?
+  - What prevents them from noticing that they should ask for help?
+  - What prevents them from reaching out for help when they notice they’re stuck?
+
+### Unpredictable / highly variable velocity (typically mid-level)
+
+When a developer’s productivity fluctuates dramatically, this is often a symptom of taking on highly complex work relative to their level.
```
Comment (Jordanjc3) [inline productivity.md:61]: While I agree, I'm not sure that this is a sufficient explanation. 

Mid-level Engineers can still struggle with estimation, those lessons also apply in this section.

File: productivity.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+- Learning complex problem-solving takes time; keep at it. Unfortunately, the long term solution to this problem is to just keep practicing until you gain mastery over new skills. Keep challenging yourself, just make sure you don’t slip into consistently low velocity.
```
Comment (Jordanjc3) [inline productivity.md:63]: Agreed that this is a tough skill to learn, that said: we should be able to challenge ourselves and solve complex problems without observing variable velocity or consistently low sprints. 

If we're going to challenge ourselves, it's **more** important to do more thorough planning with the rest of the team to discuss the approach, identify the areas of uncertainty, and to create the correct set of tickets. 

The second bullet point here is the most important of the three.

File: productivity.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+- Pair when planning for tickets outside your expertise. Talk through your proposed solutions with another engineer proficient in that area to make sure your plan of attack is sound before devoting too much time to it.
+  - The same applies when you get stuck on a problem with multiple possible solutions.
+
+### Bus-factor 1 (typically senior)
+
+What would happen to the team if you got hit by a bus? When a developer is the only person who is capable of performing a task, this is a symptom of knowledge siloing.
+
+- **Document everything.** Create documentation explaining how a system or process works.
+- **Cross-train other developers.** Stop taking tickets related to siloed knowledge, and instead devote time to pairing with other developers until they are confident to maintain it themselves. Use this to drive improvements to documentation.
```
Comment (Jordanjc3) [inline productivity.md:72]: To expand this point, I like to think of this as "ownership rotations". That is, for a service that you'd like another developer to become proficient in maintaining, decide together that the developer will take the majority of incoming tickets related to that service for the next n-months. At the end of the rotation, the developer typically has a strong grasp of the service.

After the rotation is over, pick another service and do it again.

**Review by Jordanjc3 — CHANGES_REQUESTED** [review]
Excellent work on this doc so far! Just a handful of comments that I found on my first pass.

File: productivity.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+1. P1 or P2 tickets
+2. Open carryover tickets
+3. Open late-add tickets
+4. Open sprint goal tickets
+5. Open other tickets
+6. Unstarted carryover tickets
+7. Unstarted late-add tickets
+8. Unstarted sprint goal tickets
+9. Unstarted other tickets
```
[the remark this answers is not in the document]
Comment (dbhynds) [reply productivity.md:21]: The three things would be:

- P1/P2
- Open tickets
- Unstarted tickets

But within [open|unstarted] tickets, the prioirity should be:

- carryover
- late-add
- sprint goals
- other

My goal here was to provide a comprehensive list that they could put every ticket on. The problem is I often see developers making mistakes in both sets. I often see people start new tickets when they have open tickets that can be worked on. I also see them taking on low-priority tickets when they have crucial ones that should be worked on instead.

Do you think just breaking the list up would help?

File: productivity.md (hunk 1 above)
Comment (dbhynds) [reply productivity.md:61]: In my mind, estimation is a team a team-level problem, not an individual problem. Tho, I do think we can expand on this. Aside from "taking on highly complex work relative to their level," what do you think EN2s struggle with during the transition between EN1 and EN3?

File: productivity.md (hunk 3 above)
Comment (dbhynds) [reply productivity.md:72]: Redid this point and the next.

File: productivity.md (hunk 2 above)
Comment (dbhynds) [reply productivity.md:63]: Do you think there is a way to clarify or better emphasize this point?

File: productivity.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+### Accurately planning work
+
+- Talk to the product team early and often. Understanding the requirements is key to planning work accurately.
+- Collaborate with the product team if you have ideas for how to improve and/or simplify the work. They are aligned to our goal of delivering high-quality work quickly, so suggestions are welcomed.
+- Some work is very difficult to plan by just looking at it. Make prototypes to help gain a better understanding and identify potential issues.
+- Ask: What are some non-obvious interactions that might cause problems?
+- In planning meetings, get everyone to speak, so everyone is able to share their unique perspective. There might be long periods of silence while folks are thinking, and that’s ok.
+
+### Handling grossly misestimated tickets
```
[the remark this answers is not in the document]
Comment (dbhynds) [reply productivity.md:31]: Added an item about reflecting on why we missed it and estimating better.

In general, I view the "Accurately planning work" section as the proactive approach to estimation, and this section as the reactive approach. But you're right, the last step of the reactive process should feed back to preventative / proactive steps.

**Review by lin-brian-l — COMMENTED** [review]
I thought this looks good to me, but I know @Jordanjc3 has some feedback which hasn't been finalized yet; I'll let him submit final approval for this PR.

File: productivity.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jordanjc3) [reply productivity.md:21]: Thanks for breaking that down a bit further. I do think having it broken out into two sections would help. 

A flow chart would be even better, but I'm OK with putting that out-of-scope for this PR :)

File: productivity.md (hunk 1 above)
Comment (Jordanjc3) [reply productivity.md:63]: In terms of team-level vs. individual problem, I think it's both. At the end of the day, the Engineer taking on the ticket has to take responsibility for the estimate -- if Engineers don't ask "what's the right estimate of this ticket _for me_?" that can lead to estimation issues. 

To your point around what EN2s struggle with, highly variable velocity tells me some combination of the following are true: 

- Tickets are being estimated poorly. 
- Our time is going elsewhere besides the tickets initially our sprint (late-adds, meetings, etc.)
- Complexity is too high

You touch on the complexity piece here and we covered estimation, I just might add a note on the second bullet.

File: productivity.md (hunk 2 above)
Comment (Jordanjc3) [reply productivity.md:65]: My recommendation here is just to put your last bullet point at the top of the list.

**Review by Jordanjc3 — APPROVED** [review]
I have some lingering comments, but I'll leave implementing them up to you.

**Review by lin-brian-l — APPROVED** [review]
@dbhynds the flowchart is great, nice work!

**Pull request merged**